documentation/CAMARA-Security-Interoperability.md

```diff
@@ -0,0 +1,188 @@
+
+CAMARA RECOMMENDS PKCE for CSRF protection. 
+CAMARA Authorization Servers MUST implement PKCE. If PKCE is not used by the Client then the CAMARA AZ must handle **state** and **nonce** as defined in OAuth2.
+
+CAMARA Clients SHOULD use PKCE.
+
+If nonce for CSRF-protection is used then implementers must ensure that sufficient entropy is present in the nonce value.
+Please see [OAuth 2.0 Security Best Current Practice](https://oauthstuff.github.io/draft-ietf-oauth-security-topics/draft-ietf-oauth-security-topics.html#name-protecting-redirect-based-f).
+
+## Client-Initiated Backchannel Authentication Flow
+
+The [Client-Initiated Backchannel Authentication (CIBA)](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html) flow is used to initiate an out-of-band authentication of the end-user.
+
+Communication with the Backchannel Authentication Endpoint MUST utilize TLS.
+
+CIBA allows the Client to get the authentication result in three ways: poll, ping, or push. This profile allows clients to use the *poll* mode. In the Poll mode, the authentication result is retrieved by the Client by polling the OP's token endpoint using the new grant type. 
+
+### Authentication Request
+
+This CAMARA document makes the login_hint parameter REQUIRED if CIBA is used in a CAMARA context.
+
+This CAMARA document clarifies the values used in login_hint in the following way:
+
+    * **_tel_**
+
+      For phone numbers. The `login_hint` must be a tel URI as defined in [RFC 3966](https://www.rfc-editor.org/info/rfc3966) for global phone numbers without visual separators in [E.164](https://www.itu.int/rec/T-REC-E.164-201011-I/en) format. For example, `tel:+34666666666`.
+
+    * **_ipport_**
+
+      For IPv4 and IPv6 addresses, that can optionally include a port. For example, `ipport:80.90.34.2:16790`, `ipport:80.90.34.2`, `ipport:[2001:db8::1]:8080` or `ipport:[2001:db8::1]`.
+
+
+The client MUST authenticate with the authorization server as described in [Client Authentication Section](#client-authentication).
+
+## Client Credentials Flow
+
+The [OAuth 2.0 Client Credentials](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4) grant type is used to obtain a 2-legged Access Token that does not represent a user. This grant type can only be used when no personal user data is processed, and it is only a valid option to access the CAMARA APIs for these specific scenarios.
+
+
+### Access Token Request
+
+This CAMARA document makes the scope parameter REQUIRED for the OAuth2 Client Credentials Grant.
+
+The client MUST authenticate with the authorization server as described in [Client Authentication Section](#client-authentication).
+
+## The Scope Parameter
+
+Scope values determine the specific CAMARA services being requested by the Service Provider, subject to the SP being registered to use those services. The scope values must be documented in the API OAS files by all Camara API subprojects.
+
```

Comment:
Please add a statement about the general handling of an access token and scope:

Access tokens are intended to convey information about permitted scopes to the resource server and therefor must be shared with the resource servers (either as opaque tokens or self contained jwt tokens).

Reply:
Isn't that the whole purpose of access tokens for API access at a resource servers? The whole purpose of OAuth2 and OIDC is getting access tokens. 

I am astonished that somebody might see this differently. 

That sounds a little like the interpretation regarding scope where OIDC defined that the value "openid" is a MUST and some telco interpreted the "unspefied behaviour" if "openid" is missing that "openid" is not a MUST.

Back to access tokens: @Elisabeth-Ericsson do we need this? Seems obvious to me but please enlighten me

Reply:
CAMARA does not define internal API implementations, and so should stick to specifying interfaces and interface behaviour, and not try to second guess API provider implementations. Provider implementations should be treated as a "black box" and behaviour that cannot be tested externally should not be specified.

Feel free to write as many "best practice" documents as you like, but CAMARA needs to be clear as to what is a "standard" (i.e. a requirement) and what is only guidance. If a provider implementation is insecure, that is an issue for them and their internal security team.

Reply:
I agree with @eric-murray (under the assumption that I understood him correctly):

Suppose we have gateway/aggregator and behind that 2 telcos t1 and t1. From a client PoV the gateway might implement both the AZ and the RS for an API which "behind the scenes" is implemented by t1 and t2 by protocol p1 and p2 which are not necessarily the same as the Camara API. 
Not that I would recommend such a scenario but I think this happens t1 having their technical debt and t2 as well.
This setup might be secure - who knows. The access token might be exchanged to internal access tokens. Who knows. There might be performance reasons for this token transformation.

I think Camara should probably not care.

@Elisabeth-Ericsson resolve this conversation?

Reply:
Camara defines APIs and protocols to call APIs. There will be situations where L1 APIs are delivered by vendors as well as service exposure platforms. L1 service implementations require the information about access tokens. Therefore it must be assured that this information is sent to the L1 services. 
I recommend to add the hint, I proposed initially. This hint does not talk about a physical format or a potential token exchange. Proposed text is:
_Access tokens are intended to convey information about permitted scopes to the resource server and therefor must be shared with the resource servers (eventually as opaque tokens or self contained jwt tokens)._

Reply:
@Elisabeth-Ericsson
If the proposed text contains a "must", then it is a requirement, not a "hint". And if it is a requirement, then the specification needs to be very clear as to what a "resource server" is. With the current definition, some implementations will not have an identifiable "resource server".

Reply:
Oh sorry, my mistake. You are perfectly right.  I should have stated: _should be shared_

Proposed text is:
Access tokens are intended to convey information about permitted scopes to the resource server and therefor should be shared with the resource servers (eventually as opaque tokens or self contained jwt tokens).

Reply:
@eric-murray: would you be fine with the phrasing: 
Access tokens are intended to convey information about permitted scopes to the resource server respectively the API implementations and therefor should be shared with them (eventually as opaque tokens or self contained jwt tokens) such that they can be validated and evaluated in the API logic.

Reply:
Thank you for your close review. I fixed the type'o.

Reply:
I think that the whole purpose of OAuth2/OIDC access tokens is to be send to API consumers to be validated and evaluated. So everybody implementing an OAuth2 RS should know.
There are cases where the actual API implementation predates OAuth2 and the access token is consumed by what the client sees as RS but the real API consumer - the implementation - is some SOAP thingy. That is not so uncommon in the telco environment. New implementations should be OAuth2 of course.

From a Camara client perspective the access token MUST be send to the RS. Whether the access token ends up at the _real_ implementation is a different matter.

If you think this implementation advice is helpful, maybe it should be in a different document?

Reply:
I added a note about this to the document. Hope that resolves this conversation. @Elisabeth-Ericsson